### PR TITLE
All system indices are hidden indices

### PR DESCRIPTION
**Pull request by williamrandolph**
System indices should be hidden from users. Since they are already restricted indices, a users that can't view restricted indices already can't see or access them, but they should also be hidden for superusers or users that are otherwise granted advanced privileges.

To the greatest degree possible, we apply hidden settings in the transport layer, so that the system can create an index or alias that is set to visible, for example, when operating in a mixed cluster mode. However, in the case of aliases created by templates, we hide the alias in the service layer.

This change has broken a number of tests that were relaying unnecessarily on wildcard searches. In general, the fix for these issues was to apply `expand_wildcards=open,hidden` to the request.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/template/post/TransportSimulateIndexTemplateAction.java [hunk 1]
```diff
@@ -217,7 +217,11 @@ public static Template resolveTemplate(
                 // the context is only used for validation so it's fine to pass fake values for the
                 // shard id and the current timestamp
                 tempIndexService.newSearchExecutionContext(0, 0, null, () -> 0L, null, emptyMap()),
-                tempIndexService.dateMathExpressionResolverAt()
+                tempIndexService.dateMathExpressionResolverAt(),
+                // we don't have a system indices object here, so we assume
+                // the alias is not a system name.
+                // TODO: any way to get system names here?
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/action/admin/indices/template/post/TransportSimulateIndexTemplateAction.java:223]: Note to reviewer: does this need to be fixed? If so, can we do it in a follow-up?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 2]
```diff
@@ -1359,7 +1359,9 @@ private static void validateCompositeTemplate(
                 // shard id and the current timestamp
                 xContentRegistry,
                 tempIndexService.newSearchExecutionContext(0, 0, null, () -> 0L, null, emptyMap()),
-                tempIndexService.dateMathExpressionResolverAt(System.currentTimeMillis())
+                tempIndexService.dateMathExpressionResolverAt(System.currentTimeMillis()),
+                // TODO: any way to get system names here?
+                alias -> false
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1364]: Note to reviewer: does this need to be fixed? It seems like if we don't catch system names in the validation stage, we may end up catching them elsewhere. If we do need to fix this, can we do it in a follow-up?

File: server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java [hunk 3]
```diff
@@ -112,6 +198,38 @@ private void doCreateTest(Runnable runnable, String concreteIndex) {
...
+            .indices()
+            .getAliases(new GetAliasesRequest().indicesOptions(IndicesOptions.strictExpandHidden()))
+            .actionGet();
+
+        assertThat(getAliasesResponse.getAliases().size(), equalTo(1));
+        assertThat(getAliasesResponse.getAliases().get(concreteIndex).size(), equalTo(1));
+        assertThat(getAliasesResponse.getAliases().get(concreteIndex).get(0).isHidden(), equalTo(true));
+    }
+
+    private void assertAliasesAddedByTemplate(Set<String> aliasNames) throws InterruptedException, java.util.concurrent.ExecutionException {
```
Comment (gwbrown) [inline server/src/internalClusterTest/java/org/elasticsearch/action/admin/indices/create/CreateSystemIndicesIT.java:218]: Nit: This method is confusingly named - I spent a while trying to figure out why `".test-index"` was in the set when this gets called. Maybe `assertHasAliases`?

**Review by AthenaEryma — COMMENTED** [review]
Left some relatively minor comments - I do think we should plumb in the `SystemIndices` given it looks fairly easy to do.

Quite good work! I'm sure this was quite a pain tracking down all these places across the codebase. I'm only being nitpicky because I want to be sure we get this right.

File: server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterInfoServiceIT.java [hunk 4]
```diff
@@ -143,7 +143,6 @@ public void testClusterInfoServiceCollectsInformation() {
                 Settings.builder()
                     .put(Store.INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), 0)
                     .put(EnableAllocationDecider.INDEX_ROUTING_REBALANCE_ENABLE_SETTING.getKey(), EnableAllocationDecider.Rebalance.NONE)
-                    .put(IndexMetadata.SETTING_INDEX_HIDDEN, randomBoolean())
```
[the remark this answers is not in the document]
Comment (williamrandolph) [reply server/src/internalClusterTest/java/org/elasticsearch/cluster/ClusterInfoServiceIT.java:146]: This was causing a failure when creating a system index. I didn't look closely enough to see that we only create the system index sometimes. I'll put it back in safely.

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java [hunk 5]
```diff
@@ -107,13 +111,26 @@ protected void masterOperation(
             }
         }
 
+        if (isSystemIndex) {
+            if (Objects.isNull(request.settings())) {
+                request.settings(SystemIndexDescriptor.DEFAULT_SETTINGS);
+            } else if (false == request.settings().hasValue(SETTING_INDEX_HIDDEN)) {
+                request.settings(Settings.builder().put(request.settings()).put(SETTING_INDEX_HIDDEN, true).build());
+            } else if ("false".equals(request.settings().get(SETTING_INDEX_HIDDEN))) {
```
[the remark this answers is not in the document]
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:119]: Good catch -- thanks!

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 6]
```diff
@@ -2299,10 +2300,20 @@ public void testSingleSystemIndexResolutionAllowed() {
         assertThat(indexNames, containsInAnyOrder(".ml-meta"));
     }
 
+    public void testFullWildcardSystemIndicesAreHidden() {
+        threadContext.putHeader(SYSTEM_INDEX_ACCESS_CONTROL_HEADER_KEY, Boolean.FALSE.toString());
```
[the remark this answers is not in the document]
Comment (williamrandolph) [reply server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:2304]: It's not important for this test, but it affects visibility in some other ones. I'll remove it where it doesn't affect results.

File: server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java [hunk 7]
```diff
@@ -2952,7 +2963,7 @@ public void testRemoteIndex() {
     }
 
     private ClusterState systemIndexTestClusterState() {
-        Settings settings = Settings.builder().build();
+        Settings settings = Settings.builder().put("index.hidden", "true").build();
```
[the remark this answers is not in the document]
Comment (williamrandolph) [reply server/src/test/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolverTests.java:2965]: Good call, will do.

File: modules/reindex/build.gradle [hunk 8]
```diff
@@ -159,6 +159,17 @@ if (Os.isFamily(Os.FAMILY_WINDOWS)) {
...
+
+    // these tests are all relying on a call to refresh all indices, when they could easily be changed
+    // in 7.x to call the specific index they want to refresh.
+    task.skipTest("delete_by_query/70_throttle/Rethrottle to -1 which turns off throttling", "test relies on system index being non-hidden")
+    task.skipTest("delete_by_query/80_slices/Multiple slices with rethrottle", "test relies on system index being non-hidden")
+    task.skipTest("delete_by_query/80_slices/Multiple slices with wait_for_completion=false", "test relies on system index being non-hidden")
+    task.skipTest("reindex/80_slices/Multiple slices with rethrottle", "test relies on system index being non-hidden")
+    task.skipTest("reindex/80_slices/Multiple slices with wait_for_completion=false", "test relies on system index being non-hidden")
+    task.skipTest("update_by_query/70_slices/Multiple slices with rethrottle", "test relies on system index being non-hidden")
+    task.skipTest("update_by_query/70_slices/Multiple slices with wait_for_completion=false", "test relies on system index being non-hidden")
```
[the remark this answers is not in the document]
Comment (gwbrown) [reply modules/reindex/build.gradle:172]: I see you opened https://github.com/elastic/elasticsearch/issues/81188, thanks!

**Review by AthenaEryma — APPROVED** [review]
LGTM, thanks for your hard work and iteration on this one! Chasing down everywhere that needs to be changed for something like this isn't easy.

**Pull request merged**